Sources/SwiftGenKit/Parsers/Strings/StringsParser.swift

```diff
@@ -38,36 +44,76 @@ public enum Strings {
+      StringsFileParser.self,
+      StringsDictFileParser.self
+    ]
+
     public init(options: [String: Any] = [:], warningHandler: Parser.MessageHandler? = nil) throws {
       self.options = try ParserOptionValues(options: options, available: Parser.allOptions)
       self.warningHandler = warningHandler
+
+      for parser in Parser.subParsers {
+        register(parser: parser)
+      }
     }
 
-    public static let defaultFilter = "[^/]\\.strings$"
     public static let allOptions: ParserOptionList = [Option.separator]
+    public static var defaultFilter: String {
+      let extensions = Parser.subParsers.flatMap { $0.extensions }.sorted().joined(separator: "|")
+      return "[^/]\\.(?i:\(extensions))$"
+    }
 
     // Localizable.strings files are generally UTF16, not UTF8!
     public func parse(path: Path, relativeTo parent: Path) throws {
+      guard let parserType = parsers[path.extension?.lowercased() ?? ""] else {
+        throw ParserError.unsupportedFileType(path: path, supported: parsers.keys.sorted())
+      }
+
       let name = path.lastComponentWithoutExtension
+      let tableExists = tables[name] != nil
+      let parser = parserType.init(options: options)
 
-      guard tables[name] == nil else {
-        throw ParserError.duplicateTable(name: name)
-      }
-      guard let data = try? path.read() else {
-        throw ParserError.failureOnLoading(path: path.string)
-      }
+      let entries: [Entry]
+      if tableExists {
+        guard parser.isAllowedToWriteToExistingTable else {
+          throw ParserError.duplicateTable(name: name)
+        }
 
-      let plist = try PropertyListSerialization.propertyList(from: data, format: nil)
-      guard let dict = plist as? [String: String] else {
-        throw ParserError.invalidFormat
+        let existingEntries = tables[name] ?? []
+        let newEntries = try parser.parseFile(at: path)
+        let mergedEntries = Array(Dictionary([existingEntries, newEntries].joined().map { ($0.key, $0) },
+                                             uniquingKeysWith: { $1 }).values)
+        entries = mergedEntries
+      } else {
+        entries = try parser.parseFile(at: path)
       }
 
-      tables[name] = try dict.map { key, translation in
-        try Entry(key: key, translation: translation, keyStructureSeparator: options[Option.separator])
+      tables[name] = entries
+    }
+
+    func register(parser: StringsFileTypeParser.Type) {
+      for ext in parser.extensions {
+        if let old = parsers[ext] {
+          warningHandler?("""
+            error: Parser \(parser) tried to register the file type '\(ext)' already \
+            registered by \(old).
+            """,
+            #file,
+            #line)
+        }
+        parsers[ext] = parser
       }
     }
   }
 }
+
+extension StringsFileTypeParser {
+  fileprivate var isAllowedToWriteToExistingTable: Bool {
+    return self is Strings.StringsDictFileParser
```

Comment:
I'm not a fan of the inversion of control here. The parser should tell if it allows to write to existing table, instead of having the parent type know which child type has that ability. So I'd have each `StringsFileTypeParser` have its own property `var isAllowedToWriteToExistingTable: Bool { get }` instead, and one have it defined to be `true` and the other to be ``false`

Though I'm not sure I understand why the Dict parser would be allowed to extend an existing table but not the Strings one? I don't think we have an absolute guarantee that the `.stringsdict` file is gonna be parsed _after_ the `.strings` file, do we? So what if the `.stringsdict` file is parsed first, and the `.strings` file is parsed next, does that mean it can't extend the entries already populated by the `.stringsdict`? I think both Strings and StringsDict parses should be able to write in each other's tables in practice, so that boolean is probably not even necessary

Reply:
> The parser should tell if it allows to write to existing table, instead of having the parent type know which child type has that ability.

Sounds reasonable, I'll add a property to the protocol and let the individual parsers implement it.

> Though I'm not sure I understand why the Dict parser would be allowed to extend an existing table but not the Strings one?

I put a question about this as a TODO in the original PR description and @djbe [answered](https://github.com/SwiftGen/SwiftGen/pull/634#issuecomment-493388455) that Foundation prefers `.stringsdict` over `.strings` if both files are using the same table and the same key. Last year I also verified this through experimentation, and then crossed off the TODO and implemented it this way.

> I don't think we have an absolute guarantee that the .stringsdict file is gonna be parsed after the .strings file, do we?

True, I didn't think of this when I originally implemented this PR. I'll make it work both ways, so that either can be parsed first or also interleaved, since the `Parser` will just iterate over the paths without grouping by `.strings` or `.stringsdict`, this should also reflect the real behavior.

> So what if the .stringsdict file is parsed first, and the .strings file is parsed next, does that mean it can't extend the entries already populated by the .stringsdict? I think both Strings and StringsDict parses should be able to write in each other's tables in practice, so that boolean is probably not even necessary

If the `.stringsdict` file is parsed first, the `.strings` file can only add new entries, but don't overwrite existing ones.

Reply:
Ah, right. Yes, values in stringsdict take precedence at runtime over strings value; I thought at first that this setting would not even allow amending the dictionary created by stringdict with _any_ new key from strings file even if they didn't exist in stringsdict… but I get it now.

Indeed for a given key, the stringsdict one should take precedence over the strings one, but either of the two could be parsed before the other still 😉

Reply:
Thinking aloud here:

- @djbe suggested a while ago to enforce strings to be parsed first and stringsdict next, overriding potential duplicates value. If that's how we do it then indeed only the StringsDictParser should allow override, being guaranteed to be run in 2nd.
- But what if the user explicitly puts the entry for the stringsdict in their config file (instead of letting SwiftGen infer it from the strings entry) and puts it before the strings file…? We have to think of that scenario too

Reply:
With the updated implementation the order is still based on the paths and not grouped by parser, but it doesn't matter, because the StringsFileParser is not able to overwrite anything the StringsDictParser wrote before. If it encounters keys that it would like to override, it will print them in a warning.